Added a UiProvider for Tree Widget

spap975 opened the pull request:
Using existing code and props from TreeWidgetControlOptions, I created a UiProvider which functions the same as the Tree Widget.  I am more than happy to reduce the size of my changes if there are better options for implementation.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/imodeljs/viewer-components-react?pullRequest=119) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/imodeljs/viewer-components-react?pullRequest=119) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/imodeljs/viewer-components-react?pullRequest=119) it.</sub>

aruniverse commented on packages/tree-widget/package.json:
```diff
@@ -62,8 +62,8 @@
     "cpx": "^1.5.0",
-    "react": "^16.8.0",
-    "react-dom": "^16.8.0",
+    "react": "^16.13.0",
```
Did you need to bump these versions?

spap975 replied:
yeah I was running into issues when testing the provider in the web-viewer-test

aruniverse replied:
I think thats because you copied the entire pkg over into that repo, id revert these changes in this file since we know it works because its being used in itwin/viewer-react

aruniverse replied:
For some reason github says theres still changes to this file, might be line endings or something.

could you run:
```
git checkout origin/master packages/tree-widget/package.json
```

aruniverse commented on packages/tree-widget/src/components/TreeUiItemsProvider.tsx:
```diff
@@ -0,0 +1,127 @@
+import { IModelConnection, Viewport } from "@bentley/imodeljs-frontend";
+import { SelectableContentDefinition } from "@bentley/ui-components";
+
+/** Provides the property grid widget to zone 9 */
```
Comment is irrelevant here, not property-grid widget, and there is non zone 9 in ui-2.0

aruniverse commented on packages/tree-widget/src/components/TreeUiItemsProvider.tsx:
```diff
@@ -0,0 +1,127 @@
+import { SelectableContentDefinition } from "@bentley/ui-components";
+
+/** Provides the property grid widget to zone 9 */
+export class TreeUiItemsProvider implements UiItemsProvider {
```
Maybe should be called `TreeWidgetUiItemsProvider`? Idk, probably too nit picky.
@calebmshafer thoughts?

aruniverse commented on packages/tree-widget/src/components/TreeUiItemsProvider.tsx:
```diff
@@ -0,0 +1,127 @@
+} from "@bentley/ui-framework";
+import React from "react";
+import { TreeWidgetComponent } from "./TreeWidgetComponent";
+import { CategoriesTreeComponent, ModelsTreeComponent, TreeWidgetControlOptions } from "../tree-widget-react";
```
Could you import these 3 from the actual module and not the barrel file?

aruniverse replied:
Pretty sure we'll hit circular import issues; or atleast we do in other pkgs when we import things from the same pkg from the barrel

aruniverse commented on packages/tree-widget/src/components/TreeUiItemsProvider.tsx:
```diff
@@ -0,0 +1,127 @@
+/** Provides the property grid widget to zone 9 */
+export class TreeUiItemsProvider implements UiItemsProvider {
+  public readonly id = "TreeUiitemsProvider";
+  public static i18n: I18N;
```
this appears to be unused, would remove

aruniverse commented on packages/tree-widget/src/components/TreeUiItemsProvider.tsx:
```diff
@@ -0,0 +1,127 @@
+        },
+      ];
+
+      if (this._additionalTrees) trees.push(...this._additionalTrees);
```
```suggestion
      if (this._additionalTrees) {
          trees.push(...this._additionalTrees);
      }
```

aruniverse commented on packages/tree-widget/src/components/TreeUiItemsProvider.tsx:
```diff
@@ -0,0 +1,127 @@
+
+      const trees: SelectableContentDefinition[] = [
+        {
+          label: UiFramework.translate("visibilityWidget.modeltree"),
```
I guess using `UiFramework.translate` is fine, but id rather we use `IModelApp.i18n.translate`, or use `TreeUiItemsProvider.i18n.translate`

aruniverse replied:
Actually, should use TreeWidget.i18n

aruniverse reviewed (commented):
Please update the pkg readme with instructions on how to add the UiProivder into a viewer. I would argue we should remove whatever documentation we have for ui-1.0.

You're missing rush changes. This should be a minor bump.

aruniverse commented:
@spap975 Take a look at https://github.com/imodeljs/imodeljs/pull/1620, Bill add support for UiProviders to add widgets in 1.0. 
@kckst8 is that worth adding in here you think? It will help teams transition from ui-1.0 -> 2.0.

aruniverse commented:
@bsteinbk could you take a pass over this pr since your our UiItemsProvider expert? :)

aruniverse commented:
> @spap975 Take a look at [imodeljs/imodeljs#1620](https://github.com/imodeljs/imodeljs/pull/1620), Bill add support for UiProviders to add widgets in 1.0.
> @kckst8 is that worth adding in here you think? It will help teams transition from ui-1.0 -> 2.0.

Update: Im going to say ignore this comment, we'll nee to bump the core ver of this pkg and i dont want to do that in this pr.

aruniverse commented on packages/tree-widget/src/components/TreeWidgetUiItemsProvider.tsx:
```diff
@@ -0,0 +1,131 @@
+
+      const trees: SelectableContentDefinition[] = [
+        {
+          label: TreeWidget.translate("visibilityWidget.modeltree"),
```
These strings are missing.... 
Could you add them to https://github.com/imodeljs/viewer-components-react/blob/master/packages/tree-widget/public/locales/en/TreeWidget.json

aruniverse replied:
Please update https://github.com/imodeljs/viewer-components-react/blob/master/packages/tree-widget/src/components/TreeWidgetControl.tsx such that it doesn't use `UiFramework.translate`, but rather `TreeWidget.translate`

aruniverse commented on packages/tree-widget/README.md:
```diff
@@ -180,3 +180,20 @@ export class TreeWidgetControl extends WidgetControl {
+
+### Using the TreeWidgetUiItemsProvider
+
+The `TreeWidgetUiItemsProvider` functions like the `TreeWidgetControl` but is instead passed through the `uiProviders` prop of the `<Viewer>` component.
```
We still need to initialize TreeWidget. So add that to the docs here, and update TreeWidget.initialize to be a lazy init, just in case it tries to get initialized more than once.

aruniverse commented on packages/tree-widget/README.md:
````diff
@@ -180,3 +180,20 @@ export class TreeWidgetControl extends WidgetControl {
+
+```ts
+<Viewer
+  authConfig={{ oidcClient: AuthorizationClient.oidcClient }}
````
i'd honestly remove all the other unused props here. take a look at the snippet i added here https://github.com/imodeljs/viewer-components-react/tree/master/packages/measure-tools#what-to-add-in-your-itwin-viewer-based-app-itwin-app-ui

aruniverse commented:
Did the tree's actually show up for you? I keep getting an infinite spinner... 
![image](https://user-images.githubusercontent.com/11051042/125133200-7cb6ad00-e0d3-11eb-800b-0807d9e8c147.png)
Might be due to how i added this into  the web-viewer-test...
How did you obtain the iModelConnection that you passed into the UiItemsProvider?
If im not mistaken this is what i saw you had when you shared your screen earlier but i dont like using `!` to assert that `UiFramework.getIModelConnection()`  is defined.

![image](https://user-images.githubusercontent.com/11051042/125133395-d323eb80-e0d3-11eb-8d2f-59be584583d9.png)

Makes me think that we should make the iModelConnection property optional for the UiItemsProvider constructor, and if its not defined get the connection directly from UiFramework in there

aruniverse commented on packages/tree-widget/src/components/TreeWidgetUiItemsProvider.tsx:
```diff
@@ -0,0 +1,152 @@
+
+      widgets.push({
+        id: "tree",
+        label: "Tree View",
```
this should be localized

aruniverse commented on packages/tree-widget/src/TreeWidget.ts:
```diff
@@ -13,12 +13,17 @@ import { UiError, getClassName } from "@bentley/ui-abstract";
    * Called by IModelApp to initialize the Tree Widget
    * @param i18n - The internationalization service created by the IModelApp.
    */
   public static async initialize(i18n: I18N): Promise<void> {
```
could you make this optional(the arg), and default to IModelApp.i18n if i18n is undefined?

aruniverse commented on packages/tree-widget/README.md:
````diff
@@ -180,3 +180,15 @@ export class TreeWidgetControl extends WidgetControl {
+```ts
+<Viewer
+  ...
+  onIModelAppInit={()=>{TreeWidget.initialize(IModelApp.i18n)}}
````
Initialize is an async call

aruniverse commented on packages/tree-widget/src/components/TreeWidgetUiItemsProvider.tsx:
```diff
@@ -0,0 +1,152 @@
+import { SelectableContentDefinition } from "@bentley/ui-components";
+import { TreeWidget } from "../TreeWidget";
+
+export interface TreeWidgetUiItemsProviderProps {
```
Redefining everything in TreeWidgetControlOptions seems unnecessary for when all you're trying to do is making the iModelConnection optional.

I __think__ this will suffice, please test it out

```ts
export interface TreeWidgetUiItemsProviderProps extends TreeWidgetControlOptions {
   iModelConnection?: IModelConnection;
}
```

spap975 replied:
I tried doing this before making the new interface and i got an error, that's why i made this one

aruniverse replied:
What was the error?

spap975 replied:
Interface 'TreeWidgetUiItemsProviderProps' incorrectly extends interface 'TreeWidgetControlOptions'.
  Types of property 'iModelConnection' are incompatible.
    Type 'IModelConnection | undefined' is not assignable to type 'IModelConnection'.
      Type 'undefined' is not assignable to type 'IModelConnection'.

aruniverse replied:
Ok guess you can only extend and make an optional prop required, not vice versa :/

Anyway, can type the arg in the constructor as `Partial<TreeWidgetControlOptions>` like so: 
![image](https://user-images.githubusercontent.com/11051042/125337811-6d2aa480-e31d-11eb-8596-10adc30ecd7d.png)

aruniverse commented on packages/tree-widget/README.md:
```diff
@@ -180,3 +180,15 @@ export class TreeWidgetControl extends WidgetControl {
+
+### Using the TreeWidgetUiItemsProvider
+
+The `TreeWidgetUiItemsProvider` functions like the `TreeWidgetControl` but is instead passed through the `uiProviders` prop of the `<Viewer>` component. `TreeWidget` must be initialized before use.
```
nit: should be `<Viewer />` not `<Viewer>`.

aruniverse commented on packages/tree-widget/README.md:
````diff
@@ -180,3 +180,15 @@ export class TreeWidgetControl extends WidgetControl {
 }
 ```
+
+### Using the TreeWidgetUiItemsProvider
````
@calebmshafer @kckst8 

How do we feel about this? Should this be moved up to above the readme section for ui-1.0? Do we want to remove the old way from the readme?

We need to start standardizing how we document these Extensions/UiProivders

kckst8 replied:
I think we remove the 1.0 example entirely. Nobody should be using that any longer

aruniverse reviewed (commented):
looks good. Still waiting to hear back about if this actually worked for you or if you just got an infinite spinner

spap975 commented:
> Did the tree's actually show up for you? I keep getting an infinite spinner...
> ![image](https://user-images.githubusercontent.com/11051042/125133200-7cb6ad00-e0d3-11eb-800b-0807d9e8c147.png)
> Might be due to how i added this into the web-viewer-test...
> How did you obtain the iModelConnection that you passed into the UiItemsProvider?
> If im not mistaken this is what i saw you had when you shared your screen earlier but i dont like using `!` to assert that `UiFramework.getIModelConnection()` is defined.
> 
> ![image](https://user-images.githubusercontent.com/11051042/125133395-d323eb80-e0d3-11eb-8d2f-59be584583d9.png)
> 
> Makes me think that we should make the iModelConnection property optional for the UiItemsProvider constructor, and if its not defined get the connection directly from UiFramework in there



> Did the tree's actually show up for you? I keep getting an infinite spinner...
> ![image](https://user-images.githubusercontent.com/11051042/125133200-7cb6ad00-e0d3-11eb-800b-0807d9e8c147.png)
> Might be due to how i added this into the web-viewer-test...
> How did you obtain the iModelConnection that you passed into the UiItemsProvider?
> If im not mistaken this is what i saw you had when you shared your screen earlier but i dont like using `!` to assert that `UiFramework.getIModelConnection()` is defined.
> 
> ![image](https://user-images.githubusercontent.com/11051042/125133395-d323eb80-e0d3-11eb-8d2f-59be584583d9.png)
> 
> Makes me think that we should make the iModelConnection property optional for the UiItemsProvider constructor, and if its not defined get the connection directly from UiFramework in there

I switched it so the iModelConnection is loaded in the UiProvider, if the user passes a different iModelConnection through props it'll prioritize that and default to UiFramework.getIModelConnection()

kckst8 commented on packages/tree-widget/README.md:
```diff
@@ -180,3 +180,19 @@ export class TreeWidgetControl extends WidgetControl {
+
+### Using the TreeWidgetUiItemsProvider
+
+The `TreeWidgetUiItemsProvider` functions like the `TreeWidgetControl` but is instead passed through the `uiProviders` prop of the `<Viewer />` component. `TreeWidget` must be initialized before use.
```
```suggestion
The `TreeWidgetUiItemsProvider` functions like the `TreeWidgetControl` but is instead passed through the `uiProviders` prop of the iTwin Viewer's `<Viewer />` component. `TreeWidget` must be initialized before use.
```

aruniverse commented on packages/tree-widget/README.md:
```diff
@@ -2,181 +2,28 @@
 - Categories Tree
 
-The package also provides the component which underlies `TreeWidgetControl` - `TreeWidgetComponent` - which you can wrap within your own custom widget controller and pass in your own custom trees to display.
+The package also provides the widget controller class - `TreeWidgetControl` - as well as the underlying component - `TreeWidgetComponent` - which you can wrap within your own custom UiProvider and pass in your own custom trees to display.
```
tehcnically can't use `TreeWidgetControl` in a UiProvider(ui-2.0), that's more for adding directly into a FrontstageDef (ui-1.0)

I would probably just drop the mention of the Control class.
```suggestion
The package also provides the underlying component - `TreeWidgetComponent` - which you can wrap within your own custom UiProvider and pass in your own custom trees to display.
```
